client/client.cpp

```diff
@@ -2,105 +2,139 @@
-	}
+  private:
+    std::ifstream seed_data;
+    bool init_db = false;
+    int num_clients = 16;
+    float p_get = 0.5;
+
+    std::vector<float> latencies;
+
+  public:
+    ClientHandler(int argc, char *argv[]) {
+        parseArgs(argc, argv, seed_data, init_db, num_clients, p_get);
+    }
+
+    void start() {
+        if (init_db) {
+            initDB();
+        } else {
+            runThreaded();
+        }
+    }
+
+    void initDB() {
+        redisCli rd;
+        auto pipeline = rd.pipe();
+
+        // If seed data exists, initialize the db with seed data
+        if (seed_data.is_open()) {
+            std::string line;
+            while (std::getline(seed_data, line)) {
+                Operation op = getSeedOperation(line);
+                pipeline.set(op.key, op.value);
+            }
+        }
+        // If seed data does not exist, initialize db with key from 0 to KEY_MAX
+        else {
+            for (int i = 0; i < KEY_MAX; ++i) {
+                std::string value = std::to_string(rand() % VAL_MAX);
+                pipeline.set(std::to_string(i), clientEncrypt(value));
+            }
+        }
+
+        pipeline.exec();
+    }
+
+    void runThreaded() {
+        std::vector<std::thread> threads;
+        for (int i = 0; i < num_clients; i++) {
+            threads.push_back(std::thread(&ClientHandler::run, this));
+        }
+
+        // Wait for all threads to finish
+        for (std::thread &thread : threads)
+            thread.join();
+
+        getAveLatency();
+    }
+
+    void run() {
+        auto socket = std::make_shared<TSocket>(HOST_IP, HOST_PORT);
+        auto transport = std::make_shared<TBufferedTransport>(socket);
+        auto protocol = std::make_shared<TBinaryProtocol>(transport);
+        RPCClient client(protocol);
+
+        transport->open();
+
+        std::string val;
+        // If seed data exists, run the client with data
+        if (seed_data.is_open()) {
+            std::string line;
+            while (readFile(seed_data, line)) {
+                Operation op = getSeedOperation(line);
+                auto start = high_resolution_clock::now();
+                client.access(val, op);
+                auto end = high_resolution_clock::now();
+                latencies.push_back(
+                    duration_cast<microseconds>(end - start).count());
+            }
+        }
+        // If seed data does not exist, run client on random values
+        else {
+            for (int i = 0; i < 1000; ++i) {
+                Operation op = genRandOperation(p_get);
+                auto start = high_resolution_clock::now();
+                client.access(val, op);
+                auto end = high_resolution_clock::now();
+                latencies.push_back(
+                    duration_cast<microseconds>(end - start).count());
+            }
+        }
```

Comment:
I'm good with this for now. If we revisit this, here's some improvements I would suggest:
- seed data does not exist -> NUM_OPERATIONS should be set by a flag instead of hardcoded to 1000 (for loop in else clause)
- We should be able to pass a flag for seed data, AND a flag for operations. `seed_data` is a confusing name here because it's really operations imo.
- We should abstract getting the operation so that we don't duplicate the access code, calculations, etc... what I mean is do this
```cpp
while (more_operations_exist()) {
    Operation op = get_operation();
    auto start = high_resolution_clock::now();
    client.access(val, op);
    auto end = high_resolution_clock::now();
    latencies.push_back(duration_cast<microseconds>(end-start).count());
}
```
instead of having the if (seed) ... else gen_operations